toolchain/check/handle_impl.cpp

```diff
@@ -313,8 +288,18 @@ static auto BuildImplDecl(Context& context, Parse::AnyImplDeclId node_id,
   auto [self_type_node, self_inst_id] =
       context.node_stack().PopWithNodeId<Parse::NodeCategory::ImplAs>();
   auto self_type_id = context.GetTypeIdForTypeInst(self_inst_id);
+
+  // Finish processing the name, which should be empty, but might have
+  // parameters.
+  auto name_context = context.decl_name_stack().FinishImplName();
+  CARBON_CHECK(name_context.state == DeclNameStack::NameContext::State::Empty);
+
   // Pop the `impl` introducer and any `forall` parameters as a "name".
-  auto name = PopImplIntroducerAndParamsAsNameComponent(context, node_id);
+  // Stop at the `where` node, if present, to exclude it from syntactic match.
+  auto name = PopImplIntroducerAndParamsAsNameComponent(
+      context, name_context.where_operand_node_id.is_valid()
+                   ? Parse::NodeId(name_context.where_operand_node_id)
+                   : node_id);
```

Comment:
In open discussion, we talked about checking the last node kind to make sure it is a `WhereExpr`.
```suggestion
  // Find the end node of the declaration for syntactic match.
  Parse::Tree::PostorderIterator last_param_iter(node_id);
  auto node_kind = context.parse_tree().node_kind(node_id);
  CARBON_CHECK(node_kind == Parse::NodeKind::ImplDefinitionStart ||
               node_kind == Parse::NodeKind::ImplDecl);
  // Subtracting 1 since we don't want to include the final `{`, `;`.
  --last_param_iter;
  // Stop at the `where` node, if present, to exclude it from syntactic match.
  if (context.parse_tree().node_kind(*last_param_iter) ==
      Parse::NodeKind::WhereExpr &&
      name_context.where_operand_node_id.is_valid()) {
    last_param_iter = name_context.where_operand_node_id;
    --last_param_iter;
  }               
  
  // Pop the `impl` introducer and any `forall` parameters as a "name".
  auto name = PopImplIntroducerAndParamsAsNameComponent(
      context, *last_param_iter);
```

Reply:
I think the discussed check was something like:
```
  CARBON_CHECK(name_context.where_operand_node_id.is_valid() ==
                   (context.parse_tree().node_kind(constraint_node) ==
                    Parse::NodeKind::WhereExpr),
               "where_operand_node_id: {0}; constraint node kind: {1}",
               name_context.where_operand_node_id,
               context.parse_tree().node_kind(constraint_node));
```

Note, one difference with your suggestion is that it validates the `WhereExpr` when `where_operand_node_id` is set, essentially a if-and-only-if. But I've gone a slightly different approach now, because I was realizing a similar thing could be achieved through the node stack at negligible storage cost (less for non-impls).

Reply:
Was looking at this a little further, I've at least moved out the param iter logic as requested.

Given the type of `node_id` is `AnyImplDeclId`, is the `CARBON_CHECK` still needed? I was thinking on edit that it could be removed because the validation comes from the type already.